File: 10 Days of Statistics/Day 1: Interquartile Range.R
# https://www.hackerrank.com/challenges/s10-quartiles/problem

# quartiles calculated based on the first method described in the Wikipedia:
# https://en.wikipedia.org/wiki/Quartile#Method_1
# none of the 9 types available in the R function quantile are according to this methode

Q1_Q3 <- function(v_values){
    # sort vector
    v_values <- sort(v_values)
    # number of elements in vector
    len_v <- length(v_values)
    # if number of elements is even e.g. 1,2,3,4,5,6,7,8,9,10
    if(len_v%%2==0) {
        # end of lower half will be len/2 e.g. 10/2 = 5
        end_lower <- len_v/2
        # start of upper half will be end of lower half plus 1 e.g. 5+1=6
        start_upper <- end_lower + 1
    }
    # if number is uneven e.g. 1,2,3,4,5,6,7,8,9,10,11
    else {
        # middle number e.g. (11+1)/2=6
        mid <- (len_v+1)/2
        # end of lower half will be mid minus 1 e.g. 6-1=5
        end_lower = mid - 1
        # start of upper half will be mid plus 1 e.g. 6+1=7
        start_upper = mid + 1
    }
    c(
        Q1=median(v_values[1:end_lower]),
        Q3=median(v_values[start_upper:len_v])
        )  
}

con <- file("stdin", open = "r")

dataset <- readLines(con)

v_elem <- unlist(strsplit(dataset[2], split=" "))
v_elem <- as.numeric(v_elem)

v_freq <- unlist(strsplit(dataset[3], split=" "))
v_freq <- as.numeric(v_freq)

v_values <- rep(v_elem, v_freq)

Q1_Q3_values <- Q1_Q3(v_values)
Q1 <- Q1_Q3_values[1]
Q3 <- Q1_Q3_values[2]

# format to keep the zero after the decimal point
cat(format(round(Q3-Q1, digits=1), nsmall = 1))

close(con)

File: 10 Days of Statistics/Day 1: Interquartile Range.py
# https://www.hackerrank.com/challenges/s10-interquartile-range/problem

"""
Objective
In this challenge, we practice calculating the interquartile range.

Task
The interquartile range of an array is the difference between its first (Q1) and third (Q3) quartiles (i.e., Q3-Q1).
Given an array, X, of n integers and an array, F, representing the respective frequencies of X's elements, construct a data set, S, where each xi occurs at frequency fi.
Then calculate and print S's interquartile range, rounded to a scale of 1 decimal place.

Tip: Be careful to not use integer division when averaging the middle two elements for a data set with an even number of elements, and be sure to not include the median in your upper and lower data sets.
"""

# quartiles calculated based on the first method described in the Wikipedia:
# https://en.wikipedia.org/wiki/Quartile#Method_1

# numpy not allowed for this challenge so np.repeat() not an option
# import numpy as np gives ModuleNotFoundError: No module named 'numpy'

import statistics as st

# function to create the value list by multiplying the elements by their frequencies
def elem_x_freq(ls_elem,ls_freq):
    len_ls = len(ls_elem)
    ls_values = []
    for i in range(len_ls):
        ls_values += [ls_elem[i]]*ls_freq[i]
    return ls_values

# function to split a list into two, if uneven upper half will contain the extra value
def split_list(ls_val):
    half = len(ls_val)//2
    return ls_val[:half], ls_val[half:]

# function to calculate Q1 and Q3
def Q1_Q3(ls_val, Q2):
    # if list len is uneven the med/Q2 is the mid number and not an average
    # if that is the case remove that number
    if len(ls_val)%2 == 1:
        ls_val.remove(Q2)
    ls_val_Q1, ls_val_Q3 = split_list(ls_val)
    return st.median(ls_val_Q1), st.median(ls_val_Q3)

# number of elements in list computed using len()
# following variable only needed for the challenge to work but not for the script
n = input()

ls_elem = list(map(float, input().strip().split()))
ls_freq = list(map(int, input().strip().split()))

ls_values = elem_x_freq(ls_elem,ls_freq)
ls_values.sort()

Q2 = st.median(ls_values)
Q1, Q3 = Q1_Q3(ls_values, Q2)

print(Q3-Q1)
